### PR TITLE
IOTS schema

### PR DESCRIPTION
**Pull request by Sam-Jeston**
Includes:
 - preloading contracts available on disk
 - Dynamic schema generation at time of load, compiling the valid Typescript on the fly with webpack
 - Refactor Server significantly, removing the construct of ContractApis and the associated proxy
 - Updating the docker execution engine to `load` images rather than `pull`
 - Created a tool in `test/bundles/docker` to allow easy creation of test images

### CODE REVIEW

File: src/client/Client.ts [hunk 1]
```diff
@@ -115,30 +119,11 @@ export function Client (config: Config) {
...
-          }
-        }, '')
-
-      return contractClient.mutate({
-        mutation: gql`
-            mutation {
-                ${method}(${argString})
-            }
-        `
+    async callContract ({ originatorPk, contractAddress, method, methodArguments }: ContractCallInstruction) {
```
Comment (rhyslbw) [inline src/client/Client.ts:122]: 👍

File: src/core/EngineClient.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { Bundle, Contract, ContractExecutionInstruction } from '.'
+import { Bundle, Contract, ContractCallInstruction } from '.'
```
Comment (rhyslbw) [inline src/core/EngineClient.ts:1]: 👍

File: src/server/infrastructure/engine_clients/plutus/PlutusEngineClient.ts [hunk 3]
```diff
@@ -10,10 +11,14 @@ export function PlutusEngineClient (config: Config): EngineClient {
   const { executionEndpoint, networkInterface } = config
   return {
     name: 'plutus',
+    // As we load the contracts at boot time, the execution service
+    // may not yet be available, so we allow a connection buffer here
     async loadExecutable ({ contractAddress, executable }) {
-      return networkInterface.post(`${executionEndpoint}/loadSmartContract`,
-        { contractAddress, executable: executable.toString('base64') }
-      )
+      return RetryPromise.retryPromise('loadContract', () => {
```
Comment (rhyslbw) [inline src/server/infrastructure/engine_clients/plutus/PlutusEngineClient.ts:17]: 👍

**Review by rhyslbw — CHANGES_REQUESTED** [review]
1. Looks like there's only the `Server.spec` running: http://13.238.211.79:8080/blue/organizations/jenkins/smart-contract-backend/detail/feature%2Fio-ts-impl/15/pipeline/34

2. Please update root README to ensure it reflects the current state. We have a reference to Solidity to remove at least

File: src/server/index.ts [hunk 4]
```diff
@@ -13,48 +12,43 @@ const {
...
-  !OPERATION_MODE
+  !OPERATION_MODE ||
+  !CONTRACT_DIRECTORY
 ) {
   throw new Error('Required ENVs not set')
 }
 
-const networkInterface = axios.create()
+// Allow 200mb contract images
+const networkInterface = axios.create({ maxContentLength: 200000000 })
```
Comment (rhyslbw) [inline src/server/index.ts]: Happy with this. We can easily lift to config if wanting to lock it down further

Comment (rhyslbw) [inline src/server/index.ts]: Happy with this. We can easily lift to config if wanting to lock it down further

File: package.json [hunk 5]
```diff
@@ -17,9 +17,8 @@
     "start:docker-init": "docker-compose up --build",
```
Comment (rhyslbw) [inline package.json]: I think we should drop these scripts, and just rely on the standard `docker-compose` commands. The root README already says to use `docker-compose up`

**Review by rhyslbw — COMMENTED** [review]
Running `docker-compose up` I see the _server_ timeout before _execution_service_ starts up. There's also `undefined` returned in what I assume is the connection failure log. 

```
server_1             | 01:25:12 0|index  | Attempt 10/10 of request loadContract will use delay: 2020
server_1             | 01:25:13 0|index  | Attempt 10/10 of request loadContract
server_1             | 01:25:13 0|index  | Attempt 10/10 of request loadContract will use delay: 2020
server_1             | 01:25:14 0|index  | undefined
execution_service_1  | npx: installed 319 in 27.298s
execution_service_1  | 
execution_service_1  |                         -------------
execution_service_1  | 
execution_service_1  | __/\\\\\\\\\\\\\____/\\\\____________/\\\\____/\\\\\\\\\_____
execution_service_1  |  _\/\\\/////////\\\_\/\\\\\\________/\\\\\\__/\\\///////\\\___
execution_service_1  |   _\/\\\_______\/\\\_\/\\\//\\\____/\\\//\\\_\///______\//\\\__
execution_service_1  |    _\/\\\\\\\\\\\\\/__\/\\\\///\\\/\\\/_\/\\\___________/\\\/___
execution_service_1  |     _\/\\\/////////____\/\\\__\///\\\/___\/\\\________/\\\//_____
execution_service_1  |      _\/\\\_____________\/\\\____\///_____\/\\\_____/\\\//________
execution_service_1  |       _\/\\\_____________\/\\\_____________\/\\\___/\\\/___________
execution_service_1  |        _\/\\\_____________\/\\\_____________\/\\\__/\\\\\\\\\\\\\\\_
execution_service_1  |         _\///______________\///______________\///__\///////////////__
execution_service_1  | 
```

File: test/bundles/create_docker_tar/index.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+      number2: t.number,
+    })
+
+    export const add = createEndpoint<typeof addArgs, typeof t.number, t.NullC>('add', addArgs, t.number)
+  `
+
+  return res.json({ data: schema })
+})
+
+app.listen(port, () => console.log(`Mock contract listening on port ${port}!`))
```
Comment (rhyslbw) [inline test/bundles/create_docker_tar/index.js:21]:
```
execution_service_1  | 01:30:12 0|index  | &Mock contract listening on port 8080!
```
&Mock

File: src/server/index.ts [hunk 7]
```diff
@@ -13,48 +12,43 @@ const {
...
   ]]),
-  bundleFetcher: HttpTarGzBundleFetcher(networkInterface),
   pubSubClient: OPERATION_MODE === OperationMode.distributed
     ? RedisPubSubClient({ host: REDIS_HOST, port: parseInt(REDIS_PORT) })
     : MemoryPubSubClient()
-}).boot()
+})
+
+server.boot()
   .then(() => console.log('Server booted'))
```
Comment (rhyslbw) [inline src/server/index.ts]: Let's log the graphql endpoint here

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: src/server/index.ts (hunk 4 above)
Comment (Sam-Jeston) [reply src/server/index.ts]: Updated to allow ENV override

**Comment by Sam-Jeston** [comment]
I've resolved the npx install issue. By called npx it fires an `npm i`. Instead I point to pm2 directly from the Dockerfile

File: package.json [hunk 8]
```diff
@@ -13,13 +13,10 @@
     "test:coverage": "npm test && http-server coverage",
     "api-build": "tsoa swagger && tsoa routes",
     "start:local-process": "npm run build && pm2 start --name e2e-single-process dist/single_process.js",
-    "start:docker": "docker-compose up",
-    "start:docker-init": "docker-compose up --build",
+    "start:pm2": "",
```
Comment (rhyslbw) [inline package.json]: empty script

File: package.json [hunk 9]
```diff
@@ -13,13 +13,10 @@
     "test:coverage": "npm test && http-server coverage",
     "api-build": "tsoa swagger && tsoa routes",
     "start:local-process": "npm run build && pm2 start --name e2e-single-process dist/single_process.js",
```
Comment (rhyslbw) [inline package.json]: This can now simply be `start`, and `stop`

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Just a minor issue with npm scripts

File: package.json (hunk 8 above)
Comment (Sam-Jeston) [reply package.json]: oh woops!

**Review by rhyslbw — APPROVED** [review]

**Pull request merged**